Compile bug: ggml-vulkan.cpp error: use of undeclared identifier 'conv2d_f32_cm2_len'

### Git commit

Introduced in patch: 6c7a441161080551ce8a52ba32563b6295067192

```
lama.cpp/ggml/src/ggml-vulkan/ggml-vulkan.cpp:3163:71: error: use of undeclared identifier 'conv2d_f32_cm2_len'
 3163 |                 device, device->pipeline_conv2d_f32[s], "conv2d_f32", conv2d_f32_cm2_len, conv2d_f32_cm2_data, "main", 3,
```

### Operating systems

Other? (Please let us know in description)

### GGML backends

Vulkan

### Problem description & steps to reproduce

Compiling with gradle and android studio. new to this, i dont know the cmake command but it was compiling before this patch 6c7a441161080551ce8a52ba32563b6295067192

It's a lot of work to submit a proper bug report here.


```cmake
#cat ../CMakeLists.txt
cmake_minimum_required(VERSION 3.18.1)
project("llama-android")

# Set C++ standard
set(CMAKE_CXX_STANDARD 14)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

# Find packages
find_package(Threads REQUIRED)

# Configure llama.cpp build options
set(LLAMA_STATIC ON CACHE BOOL "Build static library")
set(LLAMA_BUILD_TESTS OFF CACHE BOOL "Build tests")
set(LLAMA_BUILD_EXAMPLES OFF CACHE BOOL "Build examples")
set(LLAMA_BUILD_SERVER OFF CACHE BOOL "Build server")
set(BUILD_SHARED_LIBS OFF CACHE BOOL "Build shared libraries")
set(LLAMA_CURL OFF CACHE BOOL "Disable CURL")
set(LLAMA_BUILD_COMMON ON CACHE BOOL "Build common library")

# Enable Vulkan support using LLAMA flags (correct for latest llama.cpp)
set(LLAMA_VULKAN ON CACHE BOOL "Enable Vulkan support")
set(GGML_VULKAN ON CACHE BOOL "Enable GGML Vulkan support")

# Enable Vulkan debug info and validation for debugging shader issues
set(GGML_VULKAN_DEBUG ON CACHE BOOL "Enable Vulkan debug")
set(GGML_VULKAN_VALIDATE ON CACHE BOOL "Enable Vulkan validation")
set(GGML_VULKAN_SHADER_DEBUG_INFO ON CACHE BOOL "Enable shader debug info")
set(GGML_OPENCL OFF CACHE BOOL "Disable OpenCL")
set(GGML_OPENBLAS OFF CACHE BOOL "Disable OpenBLAS for now")
set(GGML_OPENMP OFF CACHE BOOL "Disable OpenMP to avoid conflicts")
set(GGML_LLAMAFILE OFF CACHE BOOL "Disable LlamaFile")

# Log to verify LLAMA_VULKAN is set
message(STATUS "LLAMA_VULKAN: ${LLAMA_VULKAN}")
message(STATUS "GGML_VULKAN: ${GGML_VULKAN}")
message(STATUS "CMAKE_BUILD_TYPE: ${CMAKE_BUILD_TYPE}")

# Performance optimizations
set(LLAMA_NATIVE OFF CACHE BOOL "Disable native CPU optimizations for compatibility")
set(LLAMA_LTO ON CACHE BOOL "Enable link-time optimization")

# Architecture-specific optimizations
if(${ANDROID_ABI} STREQUAL "arm64-v8a")
    set(LLAMA_ARM_NEON ON CACHE BOOL "Enable NEON for ARM")
    set(LLAMA_ARM_FP16 ON CACHE BOOL "Enable FP16 for ARM")
    # Add ARM compiler flags
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -march=armv8.2-a+fp16")
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=armv8.2-a+fp16")
    message(STATUS "Building for ARM64 with FP16 support")
elseif(${ANDROID_ABI} STREQUAL "x86_64")
    # x86_64 specific flags if needed
    message(STATUS "Building for x86_64")
endif()

# Add llama.cpp as a subdirectory
add_subdirectory(llama.cpp)

# Common library is now built by llama.cpp

# Define the JNI library
add_library(llama-android SHARED
    llama_jni.cpp
)

# Find Vulkan library
find_library(vulkan-lib vulkan)

# Link libraries
target_link_libraries(llama-android
    llama
    ggml
    ${CMAKE_THREAD_LIBS_INIT}
    android
    log
)

# Link ggml-vulkan if it exists
if(TARGET ggml-vulkan)
    target_link_libraries(llama-android ggml-vulkan)
    message(STATUS "Linking ggml-vulkan backend")
endif()

# Link Vulkan if found
if(vulkan-lib)
    target_link_libraries(llama-android ${vulkan-lib})
    message(STATUS "Linking Vulkan library: ${vulkan-lib}")
endif()

# Set linker flags for 16KB page alignment
if(ANDROID)
    target_link_options(llama-android PRIVATE
        "-Wl,-z,max-page-size=16384"
    )
endif()

# Include directories
target_include_directories(llama-android PRIVATE
    ${CMAKE_SOURCE_DIR}/llama.cpp
    ${CMAKE_SOURCE_DIR}/llama.cpp/common
    ${CMAKE_SOURCE_DIR}/llama.cpp/include
    ${CMAKE_SOURCE_DIR}/llama.cpp/ggml/include
)
```

### First Bad Commit

6c7a441161080551ce8a52ba32563b6295067192

### Compile command

```shell
/Users/erg/Library/Android/sdk/cmake/3.22.1/bin/cmake \
    -H/Users/erg/foo/mobile/android/app/src/main/cpp \
    -DCMAKE_SYSTEM_NAME=Android \
    -DCMAKE_EXPORT_COMPILE_COMMANDS=ON \
    -DCMAKE_SYSTEM_VERSION=33 \
    -DANDROID_PLATFORM=android-33 \
    -DANDROID_ABI=arm64-v8a \
    -DCMAKE_ANDROID_ARCH_ABI=arm64-v8a \
    -DANDROID_NDK=/Users/erg/Library/Android/sdk/ndk/27.0.12077973 \
    -DCMAKE_ANDROID_NDK=/Users/erg/Library/Android/sdk/ndk/27.0.12077973 \
    -DCMAKE_TOOLCHAIN_FILE=/Users/erg/Library/Android/sdk/ndk/27.0.12077973/build/cmake/android.toolchain.cmake \
    -DCMAKE_MAKE_PROGRAM=/Users/erg/Library/Android/sdk/cmake/3.22.1/bin/ninja \
    -DCMAKE_CXX_FLAGS="-std=c++14 -O3 -fno-finite-math-only -funroll-loops -fomit-frame-pointer -I/opt/homebrew/include" \
    -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/Users/erg/foo/mobile/android/app/build/intermediates/cxx/Release/q5t11s4f/obj/arm64-v8a \
    -DCMAKE_RUNTIME_OUTPUT_DIRECTORY=/Users/erg/foo/mobile/android/app/build/intermediates/cxx/Release/q5t11s4f/obj/arm64-v8a \
    -B/Users/erg/foo/mobile/android/app/.cxx/Release/q5t11s4f/arm64-v8a \
    -GNinja \
    -DANDROID_STL=c++_shared \
    -DANDROID_SUPPORT_FLEXIBLE_PAGE_SIZES=ON \
    -DCMAKE_BUILD_TYPE=Release \
    -DLLAMA_VULKAN=ON \
    -DGGML_VULKAN=ON \
    -DLLAMA_NATIVE=OFF \
    -DLLAMA_LTO=OFF \
    -DLLAMA_F16C=OFF \
    -DLLAMA_AVX=OFF \
    -DLLAMA_AVX2=OFF \
    -DGGML_OPENMP=OFF \
    -DGGML_LLAMAFILE=OFF \
    -DGGML_NATIVE=OFF \
    -DGGML_AVX=OFF \
    -DGGML_AVX2=OFF \
    -DGGML_AVX512=OFF \
    -DGGML_FMA=OFF \
    -DGGML_F16C=OFF \
    -DBUILD_SHARED_LIBS=OFF \
    -DVulkan_GLSLC_EXECUTABLE=/Users/erg/Library/Android/sdk/ndk/27.0.12077973/shader-tools/darwin-x86_64/glslc \
    -DVulkan_INCLUDE_DIR=/Users/erg/Library/Android/sdk/ndk/27.0.12077973/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include
```

### Relevant log output

```shell
ccache /Users/erg/Library/Android/sdk/ndk/27.0.12077973/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++ --target=aarch64-none-linux-android33 --sysroot=/Users/erg/Library/Android/sdk/ndk/27.0.12077973/toolchains/llvm/prebuilt/darwin-x86_64/sysroot -DGGML_SCHED_MAX_COPIES=4 -DGGML_VULKAN_DEBUG -DGGML_VULKAN_SHADER_DEBUG_INFO -DGGML_VULKAN_VALIDATE -D_GNU_SOURCE -D_XOPEN_SOURCE=600 -I/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/.. -I/Users/erg/foo/mobile/android/app/.cxx/Release/q5t11s4f/arm64-v8a/llama.cpp/ggml/src/ggml-vulkan -I/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/../include -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D__BIONIC_NO_PAGE_SIZE_MACRO -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security  -std=c++14 -O3 -fno-finite-math-only -funroll-loops -fomit-frame-pointer -I/opt/homebrew/include -march=armv8.2-a+fp16 -O3 -DNDEBUG  -fPIC -Wmissing-declarations -Wmissing-noreturn -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wunreachable-code-break -Wunreachable-code-return -Wmissing-prototypes -Wextra-semi -std=gnu++17 -MD -MT llama.cpp/ggml/src/ggml-vulkan/CMakeFiles/ggml-vulkan.dir/ggml-vulkan.cpp.o -MF llama.cpp/ggml/src/ggml-vulkan/CMakeFiles/ggml-vulkan.dir/ggml-vulkan.cpp.o.d -o llama.cpp/ggml/src/ggml-vulkan/CMakeFiles/ggml-vulkan.dir/ggml-vulkan.cpp.o -c /Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/ggml-vulkan.cpp

/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/ggml-vulkan.cpp:3163:71: error: use of undeclared identifier 'conv2d_f32_cm2_len'
 3163 |                 device, device->pipeline_conv2d_f32[s], "conv2d_f32", conv2d_f32_cm2_len, conv2d_f32_cm2_data, "main", 3,
      |                                                                       ^
```

It's broken from a clean build. The previous commit to the bad one is fine for me.

I don't know if this shows what you are looking for.

```
❯ cat app/build/intermediates/cxx/Release/q5t11s4f/logs/x86_64/build_stdout_llama-android.txt
ninja: Entering directory `/Users/erg/foo/mobile/android/app/.cxx/Release/q5t11s4f/x86_64'
[0/2] Re-checking globbed directories...
[1/67] Building C object llama.cpp/ggml/src/CMakeFiles/ggml-base.dir/ggml.c.o
[2/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-base.dir/ggml.cpp.o
[3/67] Building C object llama.cpp/ggml/src/CMakeFiles/ggml-base.dir/ggml-alloc.c.o
[4/67] Building C object llama.cpp/ggml/src/CMakeFiles/ggml-base.dir/ggml-quants.c.o
[5/67] Building C object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/ggml-cpu.c.o
[6/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/hbm.cpp.o
[7/67] Building C object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/quants.c.o
[8/67] Building C object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/arch/x86/quants.c.o
[9/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-base.dir/ggml-threading.cpp.o
[10/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-base.dir/ggml-backend.cpp.o
[11/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-base.dir/ggml-opt.cpp.o
[12/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/binary-ops.cpp.o
[13/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/unary-ops.cpp.o
[14/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/amx/amx.cpp.o
[15/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/vec.cpp.o
[16/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/traits.cpp.o
[17/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-base.dir/gguf.cpp.o
[18/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/ops.cpp.o
[19/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/arch/x86/repack.cpp.o
[20/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/ggml-cpu.cpp.o
[21/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/repack.cpp.o
[22/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/amx/mmq.cpp.o
[23/67] Creating directories for 'vulkan-shaders-gen'
[24/67] No download step for 'vulkan-shaders-gen'
[25/67] Linking CXX static library llama.cpp/ggml/src/libggml-base.a
[26/67] No update step for 'vulkan-shaders-gen'
[27/67] No patch step for 'vulkan-shaders-gen'
[28/67] Linking CXX static library llama.cpp/ggml/src/libggml-cpu.a
[29/67] Performing configure step for 'vulkan-shaders-gen'
-- The C compiler identification is AppleClang 17.0.0.17000013
-- The CXX compiler identification is AppleClang 17.0.0.17000013
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found                                                                                                                                                            -- Performing Test CMAKE_HAVE_LIBC_PTHREAD                                                                                                                                                  -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success                                                                                                                                        -- Found Threads: TRUE                                                                                                                                                                      -- Enabling shader debug info                                                                                                                                                               -- Configuring done                                                                                                                                                                         -- Generating done                                                                                                                                                                          -- Build files have been written to: /Users/erg/foo/mobile/android/app/.cxx/Release/q5t11s4f/x86_64/llama.cpp/ggml/src/ggml-vulkan/vulkan-shaders-gen-prefix/src/vulkan-shaders-gen-build                                                                                                                                                                                  [30/67] Performing build step for 'vulkan-shaders-gen'                                                                                                                                      [1/2] Building CXX object CMakeFiles/vulkan-shaders-gen.dir/vulkan-shaders-gen.cpp.o                                                                                                        [2/2] Linking CXX executable /Users/erg/foo/mobile/android/app/.cxx/Release/q5t11s4f/x86_64/bin/vulkan-shaders-gen                                                             [31/67] Performing install step for 'vulkan-shaders-gen'                                                                                                                                    -- Installing: /Users/erg/foo/mobile/android/app/.cxx/Release/q5t11s4f/x86_64/Release/./vulkan-shaders-gen
[32/67] Completed 'vulkan-shaders-gen'
[33/67] Generate vulkan shaders
ggml_vulkan: Generating and compiling shaders to SPIR-V
cannot compile conv2d_f32_cm2

/Users/erg/Library/Android/sdk/ndk/27.0.12077973/shader-tools/darwin-x86_64/glslc -fshader-stage=compute --target-env=vulkan1.3 -O /Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/vulkan-shaders/conv2d_mm.comp -o /Users/erg/foo/mobile/android/app/.cxx/Release/q5t11s4f/x86_64/llama.cpp/ggml/src/ggml-vulkan/vulkan-shaders.spv/conv2d_f32_cm2.spv -g -DA_TYPE=float -DB_TYPE=float -DCOOPMAT2=1 -DD_TYPE=float -DUNROLL=[[unroll]] -DUSE_COLLECTIVES=1

/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/vulkan-shaders/conv2d_mm.comp:5: warning: '#extension' : extension not supported: GL_NV_cooperative_matrix2
/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/vulkan-shaders/conv2d_mm.comp:175: error: 'coopmat' : undeclared identifier                      /Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/vulkan-shaders/conv2d_mm.comp:175: error: '' :  syntax error, unexpected COMMA, expecting LEFT_PAREN
1 warning and 2 errors generated.

cannot compile conv2d_f16_f32_cm2
                                                                                                                                                                                            /Users/erg/Library/Android/sdk/ndk/27.0.12077973/shader-tools/darwin-x86_64/glslc -fshader-stage=compute --target-env=vulkan1.3 -O /Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/vulkan-shaders/conv2d_mm.comp -o /Users/erg/foo/mobile/android/app/.cxx/Release/q5t11s4f/x86_64/llama.cpp/ggml/src/ggml-vulkan/vulkan-shaders.spv/conv2d_f16_f32_cm2.spv -g -DA_TYPE=float16_t -DB_TYPE=float -DCOOPMAT2=1 -DD_TYPE=float -DUNROLL=[[unroll]] -DUSE_COLLECTIVES=1

/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/vulkan-shaders/conv2d_mm.comp:5: warning: '#extension' : extension not supported: GL_NV_cooperative_matrix2                                                                                                                                                                                  /Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/vulkan-shaders/conv2d_mm.comp:175: error: 'coopmat' : undeclared identifier                      /Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/vulkan-shaders/conv2d_mm.comp:175: error: '' :  syntax error, unexpected COMMA, expecting LEFT_PAREN
1 warning and 2 errors generated.
[34/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-cparams.cpp.o
[35/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-io.cpp.o                                                                                                               [36/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-arch.cpp.o
[37/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-chat.cpp.o
[38/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama.cpp.o
[39/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-impl.cpp.o
[40/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-memory-hybrid.cpp.o
[41/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-adapter.cpp.o                                                                                                          [42/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-grammar.cpp.o                                                                                                          [43/67] Building CXX object llama.cpp/ggml/src/CMakeFiles/ggml.dir/ggml-backend-reg.cpp.o
[44/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-context.cpp.o
[45/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-graph.cpp.o                                                                                                            [46/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-kv-cache-unified.cpp.o
[47/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-kv-cache-unified-iswa.cpp.o
[48/67] Building CXX object CMakeFiles/llama-android.dir/llama_jni.cpp.o                                                                                                                    /Users/erg/foo/mobile/android/app/src/main/cpp/llama_jni.cpp:488:13: warning: 'llama_kv_self_clear' is deprecated: Use llama_memory_clear() instead [-Wdeprecated-declarations]
  488 |             llama_kv_self_clear(ctx->context.get());                                                                                                                                      |             ^
/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/include/llama.h:683:5: note: 'llama_kv_self_clear' has been explicitly marked deprecated here                           683 |     DEPRECATED(LLAMA_API void llama_kv_self_clear(
      |     ^
/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/include/llama.h:29:56: note: expanded from macro 'DEPRECATED'
   29 | #    define DEPRECATED(func, hint) func __attribute__((deprecated(hint)))
      |                                                        ^                                                                                                                            /Users/erg/foo/mobile/android/app/src/main/cpp/llama_jni.cpp:885:9: warning: 'llama_kv_self_clear' is deprecated: Use llama_memory_clear() instead [-Wdeprecated-declarations]
  885 |         llama_kv_self_clear(ctx->context.get());                                                                                                                                          |         ^
/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/include/llama.h:683:5: note: 'llama_kv_self_clear' has been explicitly marked deprecated here                           683 |     DEPRECATED(LLAMA_API void llama_kv_self_clear(
      |     ^
/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/include/llama.h:29:56: note: expanded from macro 'DEPRECATED'
   29 | #    define DEPRECATED(func, hint) func __attribute__((deprecated(hint)))                                                                                                                 |                                                        ^
2 warnings generated.
[49/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-batch.cpp.o
[50/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-hparams.cpp.o
[51/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-memory.cpp.o
[52/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-mmap.cpp.o                                                                                                             [53/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-memory-recurrent.cpp.o
[54/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-model-saver.cpp.o
[55/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/unicode-data.cpp.o
[56/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-model-loader.cpp.o
[57/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-quant.cpp.o
[58/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-vocab.cpp.o
[59/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-model.cpp.o
[60/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/unicode.cpp.o
[61/67] Building CXX object llama.cpp/src/CMakeFiles/llama.dir/llama-sampling.cpp.o
[62/67] Building CXX object llama.cpp/ggml/src/ggml-vulkan/CMakeFiles/ggml-vulkan.dir/ggml-vulkan-shaders.cpp.o
[63/67] Building CXX object llama.cpp/ggml/src/ggml-vulkan/CMakeFiles/ggml-vulkan.dir/ggml-vulkan.cpp.o
FAILED: llama.cpp/ggml/src/ggml-vulkan/CMakeFiles/ggml-vulkan.dir/ggml-vulkan.cpp.o
ccache /Users/erg/Library/Android/sdk/ndk/27.0.12077973/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++ --target=x86_64-none-linux-android33 --sysroot=/Users/erg/Library/Android/sdk/ndk/27.0.12077973/toolchains/llvm/prebuilt/darwin-x86_64/sysroot -DGGML_SCHED_MAX_COPIES=4 -DGGML_VULKAN_DEBUG -DGGML_VULKAN_SHADER_DEBUG_INFO -DGGML_VULKAN_VALIDATE -D_GNU_SOURCE -D_XOPEN_SOURCE=600 -I/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/.. -I/Users/erg/foo/mobile/android/app/.cxx/Release/q5t11s4f/x86_64/llama.cpp/ggml/src/ggml-vulkan -I/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/../include -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D__BIONIC_NO_PAGE_SIZE_MACRO -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security  -std=c++14 -O3 -fno-finite-math-only -funroll-loops -fomit-frame-pointer -I/opt/homebrew/include -O3 -DNDEBUG  -fPIC -Wmissing-declarations -Wmissing-noreturn -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wunreachable-code-break -Wunreachable-code-return -Wmissing-prototypes -Wextra-semi -std=gnu++17 -MD -MT llama.cpp/ggml/src/ggml-vulkan/CMakeFiles/ggml-vulkan.dir/ggml-vulkan.cpp.o -MF llama.cpp/ggml/src/ggml-vulkan/CMakeFiles/ggml-vulkan.dir/ggml-vulkan.cpp.o.d -o llama.cpp/ggml/src/ggml-vulkan/CMakeFiles/ggml-vulkan.dir/ggml-vulkan.cpp.o -c /Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/ggml-vulkan.cpp                                                                                                                                                /Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/ggml-vulkan.cpp:3163:71: error: use of undeclared identifier 'conv2d_f32_cm2_len'
 3163 |                 device, device->pipeline_conv2d_f32[s], "conv2d_f32", conv2d_f32_cm2_len, conv2d_f32_cm2_data, "main", 3,                                                                 |                                                                       ^                                                                                                             /Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/ggml-vulkan.cpp:3163:91: error: use of undeclared identifier 'conv2d_f32_cm2_data'
 3163 |                 device, device->pipeline_conv2d_f32[s], "conv2d_f32", conv2d_f32_cm2_len, conv2d_f32_cm2_data, "main", 3,
      |                                                                                           ^
/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/ggml-vulkan.cpp:3166:79: error: use of undeclared identifier 'conv2d_f16_f32_cm2_len'; did you mean 'conv2d_f16_f32_len'?
 3166 |                 device, device->pipeline_conv2d_f16_f32[s], "conv2d_f16_f32", conv2d_f16_f32_cm2_len, conv2d_f16_f32_cm2_data, "main", 3,
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~
      |                                                                               conv2d_f16_f32_len
/Users/erg/foo/mobile/android/app/.cxx/Release/q5t11s4f/x86_64/llama.cpp/ggml/src/ggml-vulkan/ggml-vulkan-shaders.hpp:94:16: note: 'conv2d_f16_f32_len' declared here
   94 | const uint64_t conv2d_f16_f32_len = 122500;
      |                ^
/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/ggml-vulkan.cpp:3166:103: error: use of undeclared identifier 'conv2d_f16_f32_cm2_data'; did you mean 'conv2d_f16_f32_data'?
 3166 |                 device, device->pipeline_conv2d_f16_f32[s], "conv2d_f16_f32", conv2d_f16_f32_cm2_len, conv2d_f16_f32_cm2_data, "main", 3,
      |                                                                                                       ^~~~~~~~~~~~~~~~~~~~~~~
      |                                                                                                       conv2d_f16_f32_data
/Users/erg/foo/mobile/android/app/.cxx/Release/q5t11s4f/x86_64/llama.cpp/ggml/src/ggml-vulkan/ggml-vulkan-shaders.hpp:93:22: note: 'conv2d_f16_f32_data' declared here
   93 | extern unsigned char conv2d_f16_f32_data[122500];
      |                      ^
/Users/erg/foo/mobile/android/app/src/main/cpp/llama.cpp/ggml/src/ggml-vulkan/ggml-vulkan.cpp:8208:9: warning: suggest braces around initialization of subobject [-Wmissing-braces]
 8208 |         sections[0], sections[1], sections[2], sections[3], backprop
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         {                                                 }
1 warning and 4 errors generated.
ninja: build stopped: subcommand failed.

```

--- a/ggml/src/ggml-vulkan/ggml-vulkan.cpp
+++ b/ggml/src/ggml-vulkan/ggml-vulkan.cpp
@@ -3096,9 +3096,11 @@ static void ggml_vk_load_shaders(vk_device& device) {
         uint32_t conv2d_SHMEM_PAD = 4;
         bool conv2d_UNROLL = true;
 
+#if defined(GGML_VULKAN_COOPMAT2_GLSLC_SUPPORT)
         if (device->coopmat2) {
             conv2d_SHMEM_PAD = 8; // 8 float16_t
         }
+#endif
 
         if (device->vendor_id == VK_VENDOR_ID_INTEL) {
             conv2d_SHMEM_PAD = 0;
@@ -3158,6 +3160,7 @@ static void ggml_vk_load_shaders(vk_device& device) {
         std::array<uint32_t, 3> wg_denoms = { conv2d_BS_K, conv2d_BS_NPQ, 1 };
         std::vector<uint32_t> spec_constants = { conv2d_WG_SIZE, conv2d_BS_K, conv2d_BS_CRS, conv2d_BS_NPQ, conv2d_TS_K, use_collectives, conv2d_SHMEM_PAD };
 
+#if defined(GGML_VULKAN_COOPMAT2_GLSLC_SUPPORT)
         if (device->coopmat2) {
             ggml_vk_create_pipeline(
                 device, device->pipeline_conv2d_f32[s], "conv2d_f32", conv2d_f32_cm2_len, conv2d_f32_cm2_data, "main", 3,
@@ -3165,7 +3168,9 @@ static void ggml_vk_load_shaders(vk_device& device) {
             ggml_vk_create_pipeline(
                 device, device->pipeline_conv2d_f16_f32[s], "conv2d_f16_f32", conv2d_f16_f32_cm2_len, conv2d_f16_f32_cm2_data, "main", 3,
                 sizeof(vk_op_conv2d_push_constants), wg_denoms, spec_constants, 1, true, use_collectives);
-        } else if (conv2d_UNROLL) {
+        } else
+#endif
+        if (conv2d_UNROLL) {
             ggml_vk_create_pipeline(
                 device, device->pipeline_conv2d_f32[s], "conv2d_f32", conv2d_f32_unroll_len, conv2d_f32_unroll_data, "main", 3,
                 sizeof(vk_op_conv2d_push_constants), wg_denoms, spec_constants, 1, true, use_collectives);

--- a/ggml/src/ggml-vulkan/vulkan-shaders/vulkan-shaders-gen.cpp
+++ b/ggml/src/ggml-vulkan/vulkan-shaders/vulkan-shaders-gen.cpp
@@ -661,8 +661,10 @@ void process_shaders() {
     string_to_spv("conv2d_f32", "conv2d_mm.comp", {{"A_TYPE", "float"}, {"B_TYPE", "float"}, {"D_TYPE", "float"}, {"USE_COLLECTIVES", "1"}, {"UNROLL", ""}});
     string_to_spv("conv2d_f16_f32", "conv2d_mm.comp", {{"A_TYPE", "float16_t"}, {"B_TYPE", "float"}, {"D_TYPE", "float"}, {"USE_COLLECTIVES", "1"}, {"UNROLL", ""}});
 
+#if defined(GGML_VULKAN_COOPMAT2_GLSLC_SUPPORT)
     string_to_spv("conv2d_f32", "conv2d_mm.comp", {{"A_TYPE", "float"}, {"B_TYPE", "float"}, {"D_TYPE", "float"}, {"USE_COLLECTIVES", "1"}, {"UNROLL", "[[unroll]]"}, {"COOPMAT2", "1"}}, true, false, true);
     string_to_spv("conv2d_f16_f32", "conv2d_mm.comp", {{"A_TYPE", "float16_t"}, {"B_TYPE", "float"}, {"D_TYPE", "float"}, {"USE_COLLECTIVES", "1"}, {"UNROLL", "[[unroll]]"}, {"COOPMAT2", "1"}}, true, false, true);
+#endif
 
     string_to_spv("conv2d_dw_whcn_f32", "conv2d_dw.comp", merge_maps(base_dict, {{"A_TYPE", "float"}, {"B_TYPE", "float"}, {"D_TYPE", "float"}, {"WHCN", "1"}}));
     string_to_spv("conv2d_dw_cwhn_f32", "conv2d_dw.comp", merge_maps(base_dict, {{"A_TYPE", "float"}, {"B_TYPE", "float"}, {"D_TYPE", "float"}, {"CWHN", "1"}}));

